src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,320 @@
+import com.google.devtools.build.lib.authandtls.CallCredentialsProvider;
+import com.google.devtools.build.lib.concurrent.ThreadSafety.ThreadSafe;
+import com.google.devtools.build.lib.remote.RemoteRetrier.ProgressiveBackoff;
+import com.google.devtools.build.lib.remote.Retrier.Backoff;
+import com.google.devtools.build.lib.remote.common.OperationObserver;
+import com.google.devtools.build.lib.remote.common.RemoteExecutionClient;
+import com.google.devtools.build.lib.remote.options.RemoteOptions;
+import com.google.devtools.build.lib.remote.util.TracingMetadataUtils;
+import com.google.devtools.build.lib.remote.util.Utils;
+import com.google.longrunning.Operation;
+import com.google.longrunning.Operation.ResultCase;
+import com.google.rpc.Status;
+import io.grpc.Status.Code;
+import io.grpc.StatusRuntimeException;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+/**
+ * A remote work executor that uses gRPC for communicating the work, inputs and outputs.
+ *
+ * <p>It differs from {@link GrpcRemoteExecutor} by setting timeout on each execution calls to
+ * ensure we never be stuck due to network issues.
+ *
+ * @see <a href="https://docs.google.com/document/d/1NgDPsCIwprDdqC1zj0qQrh5KGK2hQTSTux1DAvi4rSc">
+ * Keepalived Remote Execution</a>
+ */
+@ThreadSafe
+public class GrpcRemoteExecutorKeepalived implements RemoteExecutionClient {
+
+  private final RemoteOptions remoteOptions;
+  private final ReferenceCountedChannel channel;
+  private final CallCredentialsProvider callCredentialsProvider;
+  private final RemoteRetrier retrier;
+
+  private final AtomicBoolean closed = new AtomicBoolean();
+
+  public GrpcRemoteExecutorKeepalived(
+      RemoteOptions remoteOptions,
+      ReferenceCountedChannel channel,
+      CallCredentialsProvider callCredentialsProvider,
+      RemoteRetrier retrier) {
+    this.remoteOptions = remoteOptions;
+    this.channel = channel;
+    this.callCredentialsProvider = callCredentialsProvider;
+    this.retrier = retrier;
+  }
+
+  private ExecutionBlockingStub executionBlockingStub() {
+    return ExecutionGrpc.newBlockingStub(channel)
+        .withInterceptors(TracingMetadataUtils.attachMetadataFromContextInterceptor())
+        .withCallCredentials(callCredentialsProvider.getCallCredentials())
+        .withDeadlineAfter(remoteOptions.remoteTimeout.getSeconds(), TimeUnit.SECONDS);
+  }
+
+  private static class Execution {
+    private final ExecuteRequest request;
+    private final OperationObserver observer;
+    private final RemoteRetrier retrier;
+    private final CallCredentialsProvider callCredentialsProvider;
+    // Count retry times for Execute() calls and can't be reset.
+    private final Backoff executeBackoff;
+    // Count retry times for WaitExecution() calls and will be reset once we received response from
```

Comment:
nit: "and is reset when we receive any response from the server that is not in error" is probably a more precise wording (although English is not my first language...)

Reply:
Done.